### PR TITLE
Adds notes about the use of libCEED

### PR DESCRIPTION
**Pull request by bishtgautam**
Adds notes on how the two `CeedOperator` are set up and how the RHSFunction works with libCEED.

### CODE REVIEW

**Comment by bishtgautam** [comment]
@jedbrown In the process of putting these notes together, I believe we can get rid of the following `VecCopy` at [physics_swe.c#L277](https://github.com/RDycore/RDycore/blob/main/src/swe/physics_swe.c#L277):

```
PetscCall(VecCopy(F, rdy->F_dup));
``` 
This can be achieved if we do a LocalToGlobal scatter on [physics_swe.c#L238](https://github.com/RDycore/RDycore/blob/main/src/swe/physics_swe.c#L238) directly to `rdy->F_dup` instead of `F` i.e. change
```
    PetscCall(DMLocalToGlobal(rdy->dm, F_local, ADD_VALUES, F));
```
to
```
    PetscCall(DMLocalToGlobal(rdy->dm, F_local, ADD_VALUES, rdy->F_dup));
```
What do you think?

File: docs/developer/ceed.md [hunk 1]
```diff
@@ -0,0 +1,206 @@
+
+# RDycore SWE physics with CEED
+
+Here we describe the RDycore's implemetation of 2D shallow water equation (SWE) that uses [libCEED](https://github.com/CEED/libCEED). The combination of PETSc and libCEED provides RDDycore algorithimic and hardware protability.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:4]: implemetation -> implementation

provides RDDycore algorithimic and hardware protability -> provides RDycore with a combination of algorithmic transparency and performance portability.

File: docs/developer/ceed.md [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+- RDycore uses PETSc's `TS` solvers that provide support for multiple time-integrators such as forward euler, RK4, etc., which can be selected at run time (e.g. `--ts_type euler`, `--ts_type rk4`, etc). 
+- The use libCEED allows RDycore to compute the RHS-function for the explicit time-integrators on CPU or GPU, which can also be selected at runtime via
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:7]: The use libCEED -> Using libCEED

the RHS-function -> source and forcing terms (?)

File: docs/developer/ceed.md [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+     - For NVIDIA GPUs: `-ceed /gpu/cuda -dm_vec_type cuda`
+     - For AMD GPUs: `-ceed /gpu/hip -dm_vec_type hip`
+
+## Example Mesh
+
+Let's consider an example 3x2 mesh that consists of
+
+- 6 cells: `c00` to `c05`
+- 17 edges: `e00` to `e16`
+    - Internal edges: `e04`, `e05`, `e07`, `e08`, `e09`, `e11`, and `e12`.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:18]: We probably don't need periods at the end of this and the next line.

File: docs/developer/ceed.md [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+- Momentum in y-dir (`hv`).
+The size of `X` will `6 * 3` where `6` corresponds to number of cells in the mesh. The layout of `X` will be as follows:
+
+```text
+X = [[h0 hu0 hv0] [h1 hu1 hv1] ... [h5 hu5 hv5]]
+```
+
+## SWE physics with libCEED
+
+The libCEED version of RDycore's explicit time-integrator of the SWE solver has two `CeedOperator`:
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:53]: `CeedOperator`s:

File: docs/developer/ceed.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+ |           |           |           |
+ |           |           |           |
+e03   c00   e04   c01   e05   c02   e06
+ |           |           |           |
+ |           |           |           |
+v00---e00---v01---e01---v02---e02---v03
+
+```
+
+For the mesh shown above, the prognostic variables of the 2D SWE are saved in a strided PETSc Vec (`X`).
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:39]: a PETSc `Vec` (`X`) with stride `3`, with components corresponding to the following prognostic variables:

File: docs/developer/ceed.md [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+```text
+X = [[h0 hu0 hv0] [h1 hu1 hv1] ... [h5 hu5 hv5]]
+```
+
+## SWE physics with libCEED
+
+The libCEED version of RDycore's explicit time-integrator of the SWE solver has two `CeedOperator`:
+
+1. `rdy->ceed_rhs.op_edges` : Computes fluxes across edges and it includes two sub-operators that correspond to:
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:55]: how about "This operator computes fluxes across edges, and includes "sub-operators" that handle internal edges and boundary edges separately."

File: docs/developer/ceed.md [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+| e05           | c01       | c02        |
+| e07           | c00       | c03        |
+| e08           | c01       | c04        |
+| e09           | c02       | c05        |
+| e11           | c03       | c04        |
+| e12           | c04       | c05        |
+
+The steps involved in creating the `CeedOperator` associated with the internal edges are as follows:
+
+- First, create a `CeedQFunction`, `qf`, and add input and output fields.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:77]: Maybe we should consider numbering these steps so we can refer to them if we need to.

Comment (jeff-cohere) [reply docs/developer/ceed.md:77]: Also, let's link to [CeedQFunction](https://libceed.org/en/latest/api/CeedQFunction/)

File: docs/developer/ceed.md (hunk 4 above)
Comment (jeff-cohere) [reply docs/developer/ceed.md:53]: Let's also link to [CeedOperator](https://libceed.org/en/latest/api/CeedOperator/) here.

File: docs/developer/ceed.md [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+The input fields includes geometric attributes associated with the edges and the prognostic variables left and right of the edges.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:78]: includes -> include
left and right of the edges -> to the left and right of the edges

File: docs/developer/ceed.md [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+| Field name | Size | In/Out | Notes                                                        |
+| ---------- | ---- | ------ | ------------------------------------------------------------ |
+| geom       |  4   | In     | Geometric attrbutes [sn, cn, L_edge/Area_left, L_edge/right] |
+| q_left     |  3   | In     | State left of the edge [h_left, hu_left, hv_left] |
+| q_right    |  3   | In     | State right of the edge [h_right, hu_right, hv_right] |
+| cell_left  |  3   | Out    | Flux contribution to the left cell [f_h f_hu f_hv] * L_edge/Area_left |
+| cell_right |  3   | Out    | Flux contribution to the right cell [f_h f_hu f_hv] * L_edge/Area_right |
+| flux       |  3   | Out    | Flux through the edge [f_h f_hu f_hv] |
+
+- Second, create `CeedElemRestriction` for all the input and output fields of previously created `CeedQFunction`.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:91]: [CeedElemRestriction](https://libceed.org/en/latest/api/CeedElemRestriction/)s

File: docs/developer/ceed.md [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+A `CeedElemRistriction` tells libCEED the indices of a `CeedVec` from/to which the values are to extracted/written
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:92]: `CeedElemRestriction`, and [CeedVector](https://libceed.org/en/latest/api/CeedVector/)

File: docs/developer/ceed.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+| `q_restrict_l`   |  `nOwnedInEdges`  | `3 * nCells`         | `CeedElemRestrictionCreate`        |  offset = [ 0,  3,  0,  3,  6,  9, 12] |
+| `q_restrict_r`   |  `nOwnedInEdges`  | `3 * nCells`         | `CeedElemRestrictionCreate`        |  offset = [ 3,  6,  9, 12, 15, 12, 15] |
+| `c_restrict_l`   |  `nOwnedInEdges`  | `3 * nCells`         | `CeedElemRestrictionCreate`        |  offset = [ 0,  3,  0,  3,  6,  9, 12] |
+| `c_restrict_r`   |  `nOwnedInEdges`  | `3 * nCells`         | `CeedElemRestrictionCreate`        |  offset = [ 3,  6,  9, 12, 15, 12, 15] |
+| `restrict_flux`  |  `nOwnedInEdges`  | `3 * nOwnedInEdges`  | `CeedElemRestrictionCreateStrided` |  offset = [ 0,  3,  6,  9, 12, 15, 18] |
+
+where `nOwnedInEdges` is the number of owned internal edges and `nCells` is the total number of cells ( = owned + ghost).
+
+- Third,  create the `CeedOperator` using the previously created `CeedQFunction` and all `CeedElementRestriction`.
+The multiple fields are added via `CeedOperatorSetField`.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:107]: [CeedOperatorSetField](https://libceed.org/en/latest/api/CeedOperator/#group__CeedOperatorUser_1gaef667c89c9ef6e043f675c800d62956b)

File: docs/developer/ceed.md [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+| e02  | c02  | d02  |
+| e03  | c00  | d03  |
+| e06  | c02  | d04  |
+| e10  | c03  | d05  |
+| e13  | c05  | d06  |
+| e14  | c03  | d07  |
+| e15  | c04  | d08  |
+| e16  | c05  | d09  |
+
+The fields for the boundary condition `CeedQFunction` are listed below. It should be noted that the `geom` field for a boundary edge has one geometric attribute less than the `geom` field for an intenral edge because the value `L_edge/A_right` is not needed for a boundary edge. Similarly, `cell_right` field is omitted for a boundary edge.
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:137]: one geometric attribute less than the `geom` field for an intenral edge -> one geometric attribute fewer than the `geom` field for an internal edge

File: docs/developer/ceed.md [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+| Field name   | CeedElemRestriction    | CeedVector           | Notes                                                                                  |
+| ------------ | ---------------------- | -------------------- | -------------------------------------------------------------------------------------- |
+| `geom`       |  `restrict_geom`       | `geom`               | `geom` has values `geom[:][0:2] = [dz/dx dz/dy]`                                       |
+| `swe_src`    |  `restrict_swe`        | `swe_src`            | `swe_src` has values for the source/sink term                                          |
+| `mannings_n` |  `restrict_mannings_n` | `mannings_n`         | `mannings_n` has values corresponding to Mannings coefficient                          |
+| `riemannf`   |  `restrict_riemannf`   | `riemannf`           | `riemannf` has the sum of Riemann fluxes in a cell through internal and boundary edges |
+| `q`          |  `restrict_q`          | `CEED_VECTOR_ACTIVE` |  |
+| `cell`       |  `restrict_c`          | `CEED_VECTOR_ACTIVE` |  |
+
+### Schematic Represntation of RHSFunction with libCEED
```
Comment (jeff-cohere) [inline docs/developer/ceed.md:204]: Represntation -> Representation

File: mkdocs.yml [hunk 14]
```diff
@@ -22,6 +22,7 @@ nav:
     - 'Code Structure and Organization': 'developer/organization.md'
     - 'Style Guide': 'developer/style.md'
     - 'Developer Tools': 'developer/tools.md'
+    - 'libCEED for SWE': 'developer/ceed.md'
```
Comment (jeff-cohere) [inline mkdocs.yml:25]: 'libCEED implementation of SWD' ?

**Review by jeff-cohere — APPROVED** [review]
Great job! This is a very clear exposition of how we use libCEED. I found some typos and suggested some rephrasings.

**Pull request merged**